xdsl/interpreter.py

```diff
@@ -274,12 +275,45 @@ def run(self, op: Operation | str):
         if isinstance(op, str):
             op = self.get_op_for_symbol(op)
 
-        inputs = self.get_values(op.operands)
         results = self._impls.run(self, op, inputs)
-        self.interpreter_assert(
-            len(op.results) == len(results), "Incorrect number of results"
-        )
-        self.set_values(zip(op.results, results))
+        return results
+
+    def run_block(self, block: Block, args: tuple[Any, ...]) -> tuple[Any, ...] | None:
+        """
+        Interpret a basic block, using `args` as the block argument values.
+        The terminator of this block is expected either to call its successor or return
+        the results for the region directly.
+        """
+        self.set_values(zip(block.args, args))
+
+        for op in block.ops:
+            inputs = self.get_values(op.operands)
+            results = self._impls.run(self, op, inputs)
+            if op.has_trait(IsTerminator):
```

Comment:
I'm not 100% sure about the use of this.
It seems that you require it (maybe because of the `toy` implementation of the interpreter?), but this is not the general condition for termination.

Would we benefit if we made the check for termination conditions a static method in the core classes and you guys can call it here while it's also being used for verification in `Operation.verify()` from `core.py`?

Reply:
I'm not really happy with it either, I'm planning on changing this. Maybe I opened up the PR too early, the final design that I have in mind is quite different to this TBH. It's still a useful first approximation but there are some subtle points that need work still

Reply:
@compor by default blocks have to end with a terminator, right?
And some operations have the NoTerminator trait otherwise, right?
If this understanding is correct, I would handle the NoTerminator case as special rather than generalize the default case, which is dominant by far, AFAIU

Reply:
Most interpretation time will be spent in this case I believe. That could be measured once up and running!

Reply:
Yeah, on your latter point, I agree that it serves its purpose here, just thinking out loud; refactoring the termination check would be trivial if we need it.

Not sure about your plan here, but I could definitely "see" modes of running/controlling the interpretation (maybe even à la debugger style, but maybe I'm also dreaming).

Reply:
If thinking about a debugger is dreaming here, you're not alone dreaming 😉

Reply:
> @compor by default blocks have to end with a terminator, right? And some operations have the NoTerminator trait otherwise, right? If this understanding is correct, I would handle the NoTerminator case as special rather than generalize the default case, which is dominant by far, AFAIU

True, it is just an exception for single-block regions; just opting out from having a terminator since control flow is rather trivial there.

Reply:
https://github.com/xdslproject/xdsl/pull/1165